contracts/farms/common/RevoFarmBot.sol

```diff
@@ -3,25 +3,38 @@ pragma solidity 0.8.4;
 
 import "hardhat/console.sol";
 
-import "./IMoolaStakingRewards.sol";
-import "./ubeswap/contracts/uniswapv2/interfaces/IUniswapV2Router02.sol";
-import "./ubeswap/contracts/uniswapv2/interfaces/IUbeswapRouter.sol";
-import "./ubeswap/contracts/uniswapv2/interfaces/IUniswapV2Pair.sol";
-import "./IRevoFees.sol";
-import "./openzeppelin-solidity/contracts/ERC20.sol";
-import "./openzeppelin-solidity/contracts/AccessControl.sol";
-import "./openzeppelin-solidity/contracts/SafeERC20.sol";
-import "./openzeppelin-solidity/contracts/Pausable.sol";
-
-contract UbeswapFarmBot is ERC20, AccessControl, Pausable {
+import "../../openzeppelin-solidity/contracts/ERC20.sol";
+import "../../openzeppelin-solidity/contracts/AccessControl.sol";
+import "../../openzeppelin-solidity/contracts/SafeERC20.sol";
+import "../../openzeppelin-solidity/contracts/Pausable.sol";
+import "../../fees/interfaces/IRevoFees.sol";
+
+/**
+ * RevoFarmBot is an abstract base class that can be used to implement Revo farm bots on top
+ * of arbitrary sources of yield. These sources of yield come from staking some ERC20 token
+ * (the "staking token") into a contract.
+ *
+ * Minimally, subclasses must implement the _deposit, _withdraw, and _claimRewards methods that
+ * deposit/withdraw an amount of the staked token into the "farm", and claim rewards. This will
+ * result in a RevoFarmBot that does not auto-compound yield; it acts only as "storage" for a
+ * user's staking token.
+ *
+ * Notably, this class does not contain a virtual "compound" function; this is because depending
+ * on the implementation, a function repsonsible for compounding the farm bot's staked position may
+ * require different arguments from the caller.
+ *
+ * This class also does not make any assumptions about how yield rewards are converted back into
+ * the staking token. For example, this may require a series of swaps and minting a liquidity token ("LP"),
+ * or for a single-staking source of yield that issues the staking token itself as a reward, nothing at all.
```

Comment:
is there an example you have in mind of a "single-staking source of yield that issues the staking token itself as a reward"? Specifically one where a farm bot would be useful?

Reply:
I worry this is over-generalized. I don't think we've ever talked about supporting this kind of use case.

Reply:
In general when there are more than 2 layers of inheritance it sets off some alarm bells for me.

Reply:
I agree, the two levels of abstraction is a little strange, though I worry that if we make too many assumptions in the base class (in particular, around the Uniswap routing/swapping protocol), we'll be setting ourselves up for a headache if/when we integrate with a platform that doesn't use Uniswap contracts. There would still be a lot of code in the base class that such an integration could and should reuse, but (at least without some more thought) wouldn't be able to, since the base class would be making too many assumptions.

You mentioned preferring composition over inheritance once before, and I absolutely agree; I looked into this a bunch while researching ways to refactor this. Composition is pretty hard in Solidity, especially when the composed units need to interact with other smart contracts. Each "object" or chunk of functionality that you're composing is a different smart contract, so you'd need to reference those from your "main" contract. That in itself is fine; the problem arises when those composed pieces of functionality need to call privileged methods on other contracts. With something like an ERC20 `transfer`, there are ways to work around this (e.g., by `approve`ing the composed contract and calling `transferFrom` from within it.). Many other contracts of course don't have the functionality to allow addresses to call methods on behalf of another address. Using libraries was my compromise; reusable chunks of code that end up getting compiled within the calling contract itself.

My vote would be to stick with this for now, and to try implementing Sushi/Ubeswap single reward farms on top of this abstraction, and see how we feel.